File: tools/ci/test_native_file_broker.js
'use strict';
const assert = require('assert');
const NativeFileBroker = require('../../plugins/robot_windows/blockly/webeeblocks/native_file_broker.js');

(async function() {
  const sent = [];
  const broker = new NativeFileBroker({send: message => sent.push(message)}, {timeoutMs: 1000});
  const pending = broker.requestCapabilities();
  assert.deepStrictEqual(sent, ['WEBEEBLOCKS_FILE_BROKER_V1 REQUEST 1 CAPABILITIES']);
  assert.strictEqual(broker.handleMessage('WEBEEBLOCKS_RUNTIME_V2 READY'), false);
  assert.strictEqual(broker.handleMessage(
    'WEBEEBLOCKS_FILE_BROKER_V1 RESPONSE 1 CAPABILITIES ' +
    '{"protocol":1,"provider":"qt6-qfiledialog-constructed","providerInjectable":true,"operationsReady":false,"canonicalExtension":".wbb"}'
  ), true);
  assert.strictEqual((await pending).operationsReady, false);
  console.log('PASS: browser capability-only broker handshake parser');
})().catch(error => { console.error(error); process.exit(1); });

File: tools/ci/test_native_file_broker.cpp
#include "file_broker.hpp"

#include <cassert>
#include <cstring>
#include <iostream>
#include <memory>
#include <string>

class InjectedProvider final : public webeeblocks::FileDialogProvider {
public:
  std::string name() const override { return "ci-injected-dialog"; }
};

int main() {
  webeeblocks::FileBroker broker(std::make_unique<InjectedProvider>());
  char response[512];
  assert(broker.handleMessage("WEBEEBLOCKS_FILE_BROKER_V1 REQUEST 7 CAPABILITIES", response, sizeof(response)));
  const std::string value(response);
  assert(value.find("RESPONSE 7 CAPABILITIES") != std::string::npos);
  assert(value.find("\"provider\":\"ci-injected-dialog\"") != std::string::npos);
  assert(value.find("\"operationsReady\":false") != std::string::npos);
  std::memset(response, 'x', sizeof(response));
  assert(!broker.handleMessage("WEBEEBLOCKS_RUNTIME_V2 REQUEST 7 RANGE front", response, sizeof(response)));
  assert(response[0] == '\0');
  assert(!broker.handleMessage("WEBEEBLOCKS_FILE_BROKER_V1 REQUEST 7 CAPABILITIES trailing", response, sizeof(response)));
  std::cout << "PASS: injectable capability-only broker contract\n";
}
